src/peft/tuners/bone/layer.py

```diff
@@ -0,0 +1,260 @@
+# Copyright 2024-present the HuggingFace Inc. team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import math
+import warnings
+from typing import Any, List, Optional, Union
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+from peft.tuners.tuners_utils import BaseTunerLayer, check_adapters_to_merge
+
+from einops import rearrange
```

Comment:
We don't have einops as a PEFT dependency. Since this is only used in one place, could you re-write it without using `rearrange`?

Reply:
I need test

Reply:
![image](https://github.com/user-attachments/assets/f32e3455-86cc-433c-b6e1-60f4d5f2e058)
Looks good.